cxx/non-atomic-action-interpreter-module/interpreter/NonAtomicActionInterpreter.cpp

```diff
@@ -0,0 +1,230 @@
+  }
+}
+
+ScAddr NonAtomicActionInterpreter::getFirstSubAction(ScAddr const & decompositionTuple)
+{
+  ScAddr firstAction =
+      utils::IteratorUtils::getAnyByOutRelation(context, decompositionTuple, scAgentsCommon::CoreKeynodes::rrel_1);
+  if (!firstAction.IsValid())
+  {
+    throw std::runtime_error("Non atomic action structure is incorrect. Failed to find first action.");
+  }
+  return firstAction;
+}
+
+void NonAtomicActionInterpreter::applyAction(ScAddr const & actionAddr)
+{
+  SC_LOG_DEBUG("NonAtomicActionInterpreter: waiting for atomic action finish.");
+  if (!utils::AgentUtils::applyAction(context, actionAddr, NonAtomicActionInterpreterConstants::INTERPRETER_ACTION_WAIT_TIME))
+  {
+    throw std::runtime_error("NonAtomicActionInterpreter: action wait time expired.");
+  }
+  SC_LOG_DEBUG("NonAtomicActionInterpreter: atomic action finished.");
+}
+
+ScAddr NonAtomicActionInterpreter::getNextAction(
+    ScAddr const & actionAddr,
+    std::map<ScAddr, ScAddr, ScAddrLessFunc> const & replacements,
+    std::map<ScAddr, std::string, ScAddrLessFunc> const & resolvedVariableIdentifiers)
+{
+  ScAddrVector orderedTransitionCandidates = getOrderedTransitionCandidates(actionAddr);
+
+  ScAddr nextAction;
+  for (auto const & transitionArc : orderedTransitionCandidates)
+  {
+    if (checkTransitionCondition(transitionArc, replacements, resolvedVariableIdentifiers))
+    {
+      nextAction = context->GetEdgeTarget(transitionArc);
+      break;
+    }
+  }
+  return nextAction;
+}
+
+ScAddrVector NonAtomicActionInterpreter::getOrderedTransitionCandidates(ScAddr const & actionAddr)
+{
+  bool actionIsSuccessful = false;
+  bool actionIsUnsuccessful = false;
+  if (context->HelperCheckEdge(
+          CoreKeynodes::question_finished_successfully, actionAddr, ScType::EdgeAccessConstPosPerm))
+  {
+    SC_LOG_DEBUG("NonAtomicActionInterpreter: atomic action finished successfully.");
+    actionIsSuccessful = true;
+  }
+  else if (context->HelperCheckEdge(
+               CoreKeynodes::question_finished_unsuccessfully, actionAddr, ScType::EdgeAccessConstPosPerm))
+  {
+    SC_LOG_DEBUG("NonAtomicActionInterpreter: atomic action finished unsuccessfully.");
+    actionIsUnsuccessful = true;
+  }
+  else
+    SC_LOG_DEBUG("NonAtomicActionInterpreter: atomic action finished with unknown result.");
+
+  ScAddrVector orderedTransitionCandidates =
+      getOrderedTransitionCandidatesFromSequence(actionAddr, actionIsSuccessful, actionIsUnsuccessful);
+
+  updateTransitionCandidatesWithTheCandidatesNotFromSequence(
+      orderedTransitionCandidates, actionAddr, actionIsSuccessful, actionIsUnsuccessful);
+
+  return orderedTransitionCandidates;
+}
+
+ScAddrVector NonAtomicActionInterpreter::getOrderedTransitionCandidatesFromSequence(
+    ScAddr const & actionAddr,
+    bool const & actionIsSuccessful,
+    bool const & actionIsUnsuccessful)
+{
+  ScAddrVector orderedTransitionCandidates;
```

Comment:
```suggestion
  ScAddrList orderedTransitionCandidates;
```
this collection is used for push_back, search and iteration, there is no access by index so it'll be better to use list